apiserver/apiserver.go

```diff
@@ -538,9 +544,33 @@ func (srv *Server) serveConn(wsConn *websocket.Conn, modelUUID string, apiObserv
 
 	conn := rpc.NewConn(codec, apiObserver)
 
-	h, err := srv.newAPIHandler(conn, modelUUID, host)
+	// Note that we don't overwrite modelUUID here because
+	// newAPIHandler treats an empty modelUUID as signifying
+	// the API version used.
+	resolvedModelUUID, err := validateModelUUID(validateArgs{
+		statePool: srv.statePool,
+		modelUUID: modelUUID,
+	})
+	var (
+		st *state.State
+		h  *apiHandler
+	)
+	if err == nil {
+		st, err = srv.statePool.Get(resolvedModelUUID)
+	}
+
+	if err == nil {
```

Comment:
The reason I merged it back in was because it felt weird to have the Get and Put in different scopes. If newAPIHandler returned an error from some reason (even though the Get succeeded) the Put would never be done, and that state would be leaked.